Add new ZoneParser API

tmthrgd opened the pull request:
This is a pretty large pull request, but again it's ultimately rather simple. The `parseZone` goroutine is eliminated and replaced with a new `ZoneParser` type with a simple chan-like API. This eliminates the goroutines and the 10,000 element buffered `parseZone` chan, which will substantially reduce memory usage.

Errors such as #786 should be impossible with the new API.

`ParseZone` maintains the same semantics as previous, including using a buffered chan of 10,000 elements. It is now a wrapper around the `ZoneParser` API.

The API (minus the `SetX` methods) was discussed in https://github.com/miekg/dns/issues/786#issuecomment-429630595.

~**Note:** This pull request contains an API change. `NewRR` and `ReadRR` no longer support the `$INCLUDE` directive. See the `SetIncludeAllowed` comment for why.~

Note: This pull request fixes a DOS vulnerability and infinite loop in processing of $GENERATE directives, see 213fc6e4e22915a8c2712ebe5a983c103fde442b.

This is the comparison between tip (17c1bc6792fdf1918200e9a675cdf2e3c9d585cd) and the pull request:
```
name                  old time/op    new time/op    delta
NewRR-12                6.57µs ± 2%    2.42µs ± 2%  -63.19%  (p=0.000 n=24+24)
ReadRR-12               6.99µs ± 1%    2.77µs ± 5%  -60.36%  (p=0.000 n=25+24)
ParseZone-12            33.9µs ±11%    40.3µs ± 9%  +18.93%  (p=0.000 n=25+23)
Generate-12              720µs ± 3%     154µs ± 3%  -78.59%  (p=0.000 n=25+25)
ParseZoneGenerate-12     720µs ± 3%     365µs ±29%  -49.29%  (p=0.000 n=25+25)

name                  old alloc/op   new alloc/op   delta
NewRR-12                  808B ± 0%      816B ± 0%   +0.99%  (p=0.000 n=25+25)
ReadRR-12               1.82kB ± 0%    1.82kB ± 0%   +0.44%  (p=0.000 n=25+25)
ParseZone-12            83.9kB ± 0%    84.0kB ± 0%   +0.19%  (p=0.000 n=25+25)
Generate-12              188kB ± 0%      32kB ± 0%  -83.09%  (p=0.000 n=25+25)
ParseZoneGenerate-12     188kB ± 0%     120kB ± 0%  -36.28%  (p=0.000 n=25+25)

name                  old allocs/op  new allocs/op  delta
NewRR-12                  17.0 ± 0%      15.0 ± 0%  -11.76%  (p=0.000 n=25+25)
ReadRR-12                 19.0 ± 0%      17.0 ± 0%  -10.53%  (p=0.000 n=25+25)
ParseZone-12              91.0 ± 0%      92.0 ± 0%   +1.10%  (p=0.000 n=25+25)
Generate-12              3.20k ± 0%     1.55k ± 0%  -51.45%  (p=0.000 n=25+25)
ParseZoneGenerate-12     3.20k ± 0%     1.69k ± 0%  -47.39%  (p=0.000 n=25+25)
```

_`ParseZoneGenerate` is `$GENERATE` used with `ParseZone` at both master and PR HEAD, while `Generate` is `$GENERATE` used with `ParseZone` at master and `ZoneParser` at PR HEAD._

This is the comparison between `ParseZone` (old) and `ZoneParser` (new)—both after this pull request:
```
name           old time/op    new time/op    delta
Parse-12       40.3µs ± 9%    10.6µs ± 1%  -73.80%  (p=0.000 n=23+24)
Generate-12     365µs ±29%     154µs ± 3%  -57.79%  (p=0.000 n=25+25)

name           old alloc/op   new alloc/op   delta
Parse-12       84.0kB ± 0%     1.6kB ± 0%  -98.10%  (p=0.000 n=25+25)
Generate-12     120kB ± 0%      32kB ± 0%  -73.46%  (p=0.000 n=25+25)

name            old allocs/op  new allocs/op  delta
Parse-12         92.0 ± 0%      81.0 ± 0%  -11.96%  (p=0.000 n=25+25)
Generate-12     1.69k ± 0%     1.55k ± 0%   -7.72%  (p=0.000 n=25+25)
```

Fixes #786
Updates coredns/coredns#2194

miekg commented:
[ Quoting <notifications@github.com> in "[miekg/dns] Add new ZoneParser API ..." ]
>This is a pretty large pull request, but again it's ultimately rather simple. The `parseZone` goroutine is eliminated and replaced with a new `ZoneParser` type with a simple chan-like API. This eliminates the goroutines and the 10,000 element buffered `parseZone` chan, which will substantially reduce memory usage.
>
>Errors such as #786 should be impossible with the new API.
>
>`ParseZone` maintains the same semantics as previous, including using a buffered chan of 10,000 elements. It is now a wrapper around the `ZoneParser` API.
>
>The API (minus the `SetX` methods) was discussed in https://github.com/miekg/dns/issues/786#issuecomment-429630595.
>
>**Note:** This pull request contains an API change. `NewRR` and `ReadRR` no longer support the `$INCLUDE` directive. See the `SetIncludeAllowed` comment for why.

This is awesome, but why the API change? Even if it's a bad idea your code will 
break subtlely (instead of not compiling). Can we reverse it, by setting to 
allow by default?

tmthrgd commented:
@miekg I reverted that part of the pull request. `ReadRR` and `NewRR` will continue allowing `$INCLUDE` directives, although I think this should be revisited in a future pull request.

I also caught and fixed an infinite loop and DOS vulnerability in the `$GENERATE` code, see 213fc6e4e22915a8c2712ebe5a983c103fde442b.

At 40 commits (coincidentally the same number as the lexer PR), I believe this is now ready to be merged. Consumers (such as coredns) can then be migrated to the new API.

tmthrgd commented:
This is what the coredns diff would look like:
```diff
diff --git a/plugin/file/file.go b/plugin/file/file.go
index 2711282ebc..4b01a3fa61 100644
--- a/plugin/file/file.go
+++ b/plugin/file/file.go
@@ -120,33 +120,31 @@ func (s *serialErr) Error() string {
 // If serial >= 0 it will reload the zone, if the SOA hasn't changed
 // it returns an error indicating nothing was read.
 func Parse(f io.Reader, origin, fileName string, serial int64) (*Zone, error) {
-	tokens := dns.ParseZone(f, dns.Fqdn(origin), fileName)
-	defer func() {
-		// Drain the tokens chan so that large zone files won't
-		// leak goroutines and memory.
-		for range tokens {
-		}
-	}()
+	zp := dns.NewZoneParser(f, dns.Fqdn(origin), fileName)
+	zp.SetIncludeAllowed(true)
+
 	z := NewZone(origin, fileName)
-	seenSOA := false
-	for x := range tokens {
-		if x.Error != nil {
-			return nil, x.Error
-		}
 
-		if !seenSOA && serial >= 0 {
-			if s, ok := x.RR.(*dns.SOA); ok {
-				if s.Serial == uint32(serial) { // same serial
+	seenSOA := false
+	for rr, ok := zp.Next(); ok; rr, ok = zp.Next() {
+		if !seenSOA {
+			if s, ok := rr.(*dns.SOA); ok {
+				if serial >= 0 && s.Serial == uint32(serial) { // same serial
 					return nil, &serialErr{err: "no change in SOA serial", origin: origin, zone: fileName, serial: serial}
 				}
 				seenSOA = true
 			}
 		}
 
-		if err := z.Insert(x.RR); err != nil {
+		if err := z.Insert(rr); err != nil {
 			return nil, err
 		}
 	}
+
+	if err := zp.Err(); err != nil {
+		return nil, err
+	}
+
 	if !seenSOA {
 		return nil, fmt.Errorf("file %q has no SOA record", fileName)
 	}
```

miekg commented on scan.go:
```diff
@@ -146,78 +157,243 @@ func ReadRR(q io.Reader, filename string) (RR, error) {
+
+	h RR_Header
+
+	sub    *ZoneParser
```
Why does zp need to carry itself? For handling includes?

tmthrgd replied:
This is for `$INCLUDE` and `$GENERATE`. They create a sub parser that’s forwarded in `subNext`.

miekg commented on scan.go:
```diff
@@ -146,78 +157,243 @@ func ReadRR(q io.Reader, filename string) (RR, error) {
+//
+// To prevent memory leaks it is important to always fully drain the
+// returned channel. If an error occurs, it will always be the last
+// Token sent on the channel.
```
Might be nice to put in a little example code.

tmthrgd replied:
I can do that, I avoided it because the `defer` block (like in the coredns PR) doesn’t `gofmt` that nicely and is a bit messy. I had thought about adding a `func DrainTokenChan(chan *Token)` that callers could `defer`.

miekg commented on scan.go:
```diff
@@ -146,78 +157,243 @@ func ReadRR(q io.Reader, filename string) (RR, error) {
+// returned channel. If an error occurs, it will always be the last
+// Token sent on the channel.
+//
+// Deprecated: New users should prefer the ZoneParser API.
```
Add maybe: See the comments in the ZoneParser API?

tmthrgd replied:
The documentation for the two is mostly the same, what was your thinking here?

miekg commented:
more tests and they pass as do the old tests. I think this is good to go - hard to properly review due to the amount of changes, trusting the tests a little here.

btw: I think the reader bufsize for NewRR can be even smaller, like 256

tmthrgd commented:
I’ll wait for your reply on the review comments above. I tried to make it easy to review, but I accept it’s a large PR.

`NewRR` isn’t buffered at all because `strings.Reader` implements `io.ByteReader` which hits the fast path in `newZLexer`.

tmthrgd commented:
> Ok, that's fine, but can you add a comment to this, so its more clear?

@miekg Will do.

> How to use the ZoneParser API, mostly?

@miekg Isn't that covered by the documentation of `ZoneParser`? It includes example code equivalent to the example code on `ParseZone`.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [miekg/dns] Add new ZoneParser ..." ]
>> Ok, that's fine, but can you add a comment to this, so its more clear?
>
>@miekg Will do.
>
>> How to use the ZoneParser API, mostly?
>
>@miekg Isn't that covered by the documentation of `ZoneParser`? It includes example code equivalent to the example code on `ParseZone`.

Then I just missed that.

/lgtm
